File: Sources/AOSRPCSchema/ComputerUse.swift
import Foundation

// MARK: - computerUse.* params / results
//
// Shell-hosted Computer Use business surface. The Sidecar exposes these as
// agent tools and calls Shell over JSON-RPC; Shell maps the wire DTOs into
// AOSComputerUseKit value types at the composition boundary.

public enum ComputerUseAppListMode: String, Codable, Sendable, Equatable {
    case running
    case all
}

public enum ComputerUseCaptureMode: String, Codable, Sendable, Equatable {
    case vision
    case ax
}

public struct ComputerUsePoint: Codable, Sendable, Equatable {
    public let x: Double
    public let y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

public struct ComputerUseBounds: Codable, Sendable, Equatable {
    public let x: Double
    public let y: Double
    public let width: Double
    public let height: Double

    public init(x: Double, y: Double, width: Double, height: Double) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

public struct ComputerUseAppInfo: Codable, Sendable, Equatable {
    public let pid: Int?
    public let bundleId: String?
    public let name: String
    public let path: String?
    public let running: Bool
    public let active: Bool
    public let identity: String

    public init(
        pid: Int?,
        bundleId: String?,
        name: String,
        path: String?,
        running: Bool,
        active: Bool,
        identity: String
    ) {
        self.pid = pid
        self.bundleId = bundleId
        self.name = name
        self.path = path
        self.running = running
        self.active = active
        self.identity = identity
    }
}

public struct ComputerUseWindowInfo: Codable, Sendable, Equatable {
    public let windowId: Int
    public let pid: Int
    public let owner: String
    public let title: String
    public let bounds: ComputerUseBounds
    public let zIndex: Int
    public let isOnScreen: Bool
    public let layer: Int

    public init(
        windowId: Int,
        pid: Int,
        owner: String,
        title: String,
        bounds: ComputerUseBounds,
        zIndex: Int,
        isOnScreen: Bool,
        layer: Int
    ) {
        self.windowId = windowId
        self.pid = pid
        self.owner = owner
        self.title = title
        self.bounds = bounds
        self.zIndex = zIndex
        self.isOnScreen = isOnScreen
        self.layer = layer
    }
}

public struct ComputerUseCoordinateSpace: Codable, Sendable, Equatable {
    public let windowFrame: ComputerUseBounds
    public let windowBounds: ComputerUseBounds
    public let pixelSize: ComputerUsePixelSize

    public init(
        windowFrame: ComputerUseBounds,
        windowBounds: ComputerUseBounds,
        pixelSize: ComputerUsePixelSize
    ) {
        self.windowFrame = windowFrame
        self.windowBounds = windowBounds
        self.pixelSize = pixelSize
    }
}

public struct ComputerUsePixelSize: Codable, Sendable, Equatable {
    public let width: Double
    public let height: Double

    public init(width: Double, height: Double) {
        self.width = width
        self.height = height
    }
}

public struct ComputerUseScreenshot: Codable, Sendable, Equatable {
    public let imageBase64: String
    public let format: String
    public let width: Int
    public let height: Int
    public let scaleFactor: Double
    public let coordinateSpace: ComputerUseCoordinateSpace
    public let originalWidth: Int?
    public let originalHeight: Int?

    public init(
        imageBase64: String,
        format: String,
        width: Int,
        height: Int,
        scaleFactor: Double,
        coordinateSpace: ComputerUseCoordinateSpace,
        originalWidth: Int?,
        originalHeight: Int?
    ) {
        self.imageBase64 = imageBase64
        self.format = format
        self.width = width
        self.height = height
        self.scaleFactor = scaleFactor
        self.coordinateSpace = coordinateSpace
        self.originalWidth = originalWidth
        self.originalHeight = originalHeight
    }
}

public struct ComputerUseListAppsParams: Codable, Sendable, Equatable {
    public let mode: ComputerUseAppListMode

    public init(mode: ComputerUseAppListMode) {
        self.mode = mode
    }
}

public struct ComputerUseListAppsResult: Codable, Sendable, Equatable {
    public let apps: [ComputerUseAppInfo]

    public init(apps: [ComputerUseAppInfo]) {
        self.apps = apps
    }
}

public struct ComputerUseListWindowsParams: Codable, Sendable, Equatable {
    public let pid: Int

    public init(pid: Int) {
        self.pid = pid
    }
}

public struct ComputerUseListWindowsResult: Codable, Sendable, Equatable {
    public let windows: [ComputerUseWindowInfo]

    public init(windows: [ComputerUseWindowInfo]) {
        self.windows = windows
    }
}

public struct ComputerUseGetAppStateParams: Codable, Sendable, Equatable {
    public let windowId: Int
    public let captureMode: ComputerUseCaptureMode
    public let maxImageDimension: Int

    public init(
        windowId: Int,
        captureMode: ComputerUseCaptureMode,
        maxImageDimension: Int
    ) {
        self.windowId = windowId
        self.captureMode = captureMode
        self.maxImageDimension = maxImageDimension
    }
}

public struct ComputerUseGetAppStateResult: Codable, Sendable, Equatable {
    public let pid: Int
    public let stateId: String
    public let treeMarkdown: String
    public let elementCount: Int
    public let screenshot: ComputerUseScreenshot?
    public let bundleId: String?
    public let appName: String?

    public init(
        pid: Int,
        stateId: String,
        treeMarkdown: String,
        elementCount: Int,
        screenshot: ComputerUseScreenshot?,
        bundleId: String?,
        appName: String?
    ) {
        self.pid = pid
        self.stateId = stateId
        self.treeMarkdown = treeMarkdown
        self.elementCount = elementCount
        self.screenshot = screenshot
        self.bundleId = bundleId
        self.appName = appName
    }
}

public struct ComputerUseStartAppSessionParams: Codable, Sendable, Equatable {
    public let pid: Int
    public let windowId: Int

    public init(pid: Int, windowId: Int) {
        self.pid = pid
        self.windowId = windowId
    }
}

public struct ComputerUseAppSessionResult: Codable, Sendable, Equatable {
    public let pid: Int

    public init(pid: Int) {
        self.pid = pid
    }
}

public struct ComputerUseStopAppSessionParams: Codable, Sendable, Equatable {
    public init() {}
}

public enum ComputerUseMouseButton: String, Codable, Sendable, Equatable {
    case left
    case right
}

public struct ComputerUseMouseClickEvent: Codable, Sendable, Equatable {
    public let button: ComputerUseMouseButton
    public let point: ComputerUsePoint
    public let count: Int

    public init(button: ComputerUseMouseButton, point: ComputerUsePoint, count: Int = 1) {
        self.button = button
        self.point = point
        self.count = count
    }
}

public struct ComputerUseMouseDragEvent: Codable, Sendable, Equatable {
    public let button: ComputerUseMouseButton
    public let from: ComputerUsePoint
    public let to: ComputerUsePoint

    public init(button: ComputerUseMouseButton, from: ComputerUsePoint, to: ComputerUsePoint) {
        self.button = button
        self.from = from
        self.to = to
    }
}

public enum ComputerUseMouseEvent: Sendable, Equatable {
    case click(ComputerUseMouseClickEvent)
    case drag(ComputerUseMouseDragEvent)
}

extension ComputerUseMouseEvent: Codable {
    private enum Kind: String, Codable {
        case click
        case drag
    }

    private enum CodingKeys: String, CodingKey {
        case kind
        case button
        case point
        case count
        case from
        case to
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch try container.decode(Kind.self, forKey: .kind) {
        case .click:
            self = .click(ComputerUseMouseClickEvent(
                button: try container.decode(ComputerUseMouseButton.self, forKey: .button),
                point: try container.decode(ComputerUsePoint.self, forKey: .point),
                count: try container.decodeIfPresent(Int.self, forKey: .count) ?? 1
            ))
        case .drag:
            self = .drag(ComputerUseMouseDragEvent(
                button: try container.decode(ComputerUseMouseButton.self, forKey: .button),
                from: try container.decode(ComputerUsePoint.self, forKey: .from),
                to: try container.decode(ComputerUsePoint.self, forKey: .to)
            ))
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .click(let event):
            try container.encode(Kind.click, forKey: .kind)
            try container.encode(event.button, forKey: .button)
            try container.encode(event.point, forKey: .point)
            try container.encode(event.count, forKey: .count)
        case .drag(let event):
            try container.encode(Kind.drag, forKey: .kind)
            try container.encode(event.button, forKey: .button)
            try container.encode(event.from, forKey: .from)
            try container.encode(event.to, forKey: .to)
        }
    }
}

public struct ComputerUsePostMouseEventParams: Codable, Sendable, Equatable {
    public let windowId: Int
    public let event: ComputerUseMouseEvent

    public init(windowId: Int, event: ComputerUseMouseEvent) {
        self.windowId = windowId
        self.event = event
    }
}

public struct ComputerUsePostMouseEventResult: Codable, Sendable, Equatable {
    public let pid: Int
    public let windowId: Int
    public let event: ComputerUseMouseEvent

    public init(pid: Int, windowId: Int, event: ComputerUseMouseEvent) {
        self.pid = pid
        self.windowId = windowId
        self.event = event
    }
}

public enum ComputerUseKeyboardModifier: String, Codable, Sendable, Equatable {
    case command
    case shift
    case option
    case control
    case function
}

public struct ComputerUseKeyboardTextEvent: Codable, Sendable, Equatable {
    public let text: String
    public let delayMilliseconds: Int

    public init(text: String, delayMilliseconds: Int = 30) {
        self.text = text
        self.delayMilliseconds = delayMilliseconds
    }
}

public struct ComputerUseKeyboardKeyPressEvent: Codable, Sendable, Equatable {
    public let key: String
    public let modifiers: [ComputerUseKeyboardModifier]
    public let count: Int

    public init(
        key: String,
        modifiers: [ComputerUseKeyboardModifier] = [],
        count: Int = 1
    ) {
        self.key = key
        self.modifiers = modifiers
        self.count = count
    }
}

public struct ComputerUseKeyboardHotkeyEvent: Codable, Sendable, Equatable {
    public let modifiers: [ComputerUseKeyboardModifier]
    public let key: String

    public init(modifiers: [ComputerUseKeyboardModifier], key: String) {
        self.modifiers = modifiers
        self.key = key
    }
}

public enum ComputerUseKeyboardEvent: Sendable, Equatable {
    case text(ComputerUseKeyboardTextEvent)
    case keyPress(ComputerUseKeyboardKeyPressEvent)
    case hotkey(ComputerUseKeyboardHotkeyEvent)
}

extension ComputerUseKeyboardEvent: Codable {
    private enum Kind: String, Codable {
        case text
        case keyPress
        case hotkey
    }

    private enum CodingKeys: String, CodingKey {
        case kind
        case text
        case delayMilliseconds
        case key
        case modifiers
        case count
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch try container.decode(Kind.self, forKey: .kind) {
        case .text:
            self = .text(ComputerUseKeyboardTextEvent(
                text: try container.decode(String.self, forKey: .text),
                delayMilliseconds: try container.decodeIfPresent(Int.self, forKey: .delayMilliseconds) ?? 30
            ))
        case .keyPress:
            self = .keyPress(ComputerUseKeyboardKeyPressEvent(
                key: try container.decode(String.self, forKey: .key),
                modifiers: try container.decodeIfPresent([ComputerUseKeyboardModifier].self, forKey: .modifiers) ?? [],
                count: try container.decodeIfPresent(Int.self, forKey: .count) ?? 1
            ))
        case .hotkey:
            self = .hotkey(ComputerUseKeyboardHotkeyEvent(
                modifiers: try container.decode([ComputerUseKeyboardModifier].self, forKey: .modifiers),
                key: try container.decode(String.self, forKey: .key)
            ))
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .text(let event):
            try container.encode(Kind.text, forKey: .kind)
            try container.encode(event.text, forKey: .text)
            try container.encode(event.delayMilliseconds, forKey: .delayMilliseconds)
        case .keyPress(let event):
            try container.encode(Kind.keyPress, forKey: .kind)
            try container.encode(event.key, forKey: .key)
            try container.encode(event.modifiers, forKey: .modifiers)
            try container.encode(event.count, forKey: .count)
        case .hotkey(let event):
            try container.encode(Kind.hotkey, forKey: .kind)
            try container.encode(event.modifiers, forKey: .modifiers)
            try container.encode(event.key, forKey: .key)
        }
    }
}

public struct ComputerUsePostKeyboardEventParams: Codable, Sendable, Equatable {
    public let windowId: Int
    public let event: ComputerUseKeyboardEvent

    public init(windowId: Int, event: ComputerUseKeyboardEvent) {
        self.windowId = windowId
        self.event = event
    }
}

public struct ComputerUsePostKeyboardEventResult: Codable, Sendable, Equatable {
    public let pid: Int
    public let windowId: Int
    public let event: ComputerUseKeyboardEvent

    public init(pid: Int, windowId: Int, event: ComputerUseKeyboardEvent) {
        self.pid = pid
        self.windowId = windowId
        self.event = event
    }
}

public enum ComputerUseAXElementAction: String, Codable, Sendable, Equatable {
    case press
    case showMenu
    case pick
    case confirm
    case cancel
    case open
    case increment
    case decrement
    case scrollToVisible
}

public enum ComputerUseAXScrollDirection: String, Codable, Sendable, Equatable {
    case up
    case down
    case left
    case right
}

public struct ComputerUseAXActionEvent: Codable, Sendable, Equatable {
    public let action: ComputerUseAXElementAction

    public init(action: ComputerUseAXElementAction) {
        self.action = action
    }
}

public struct ComputerUseAXSetValueEvent: Codable, Sendable, Equatable {
    public let value: String

    public init(value: String) {
        self.value = value
    }
}

public struct ComputerUseAXScrollEvent: Codable, Sendable, Equatable {
    public let direction: ComputerUseAXScrollDirection
    public let pages: Double

    public init(direction: ComputerUseAXScrollDirection, pages: Double) {
        self.direction = direction
        self.pages = pages
    }
}

public enum ComputerUseAXElementEvent: Sendable, Equatable {
    case action(ComputerUseAXActionEvent)
    case setValue(ComputerUseAXSetValueEvent)
    case setSelectedText(ComputerUseAXSetValueEvent)
    case focus
    case scroll(ComputerUseAXScrollEvent)
}

extension ComputerUseAXElementEvent: Codable {
    private enum Kind: String, Codable {
        case action
        case setValue
        case setSelectedText
        case focus
        case scroll
    }

    private enum CodingKeys: String, CodingKey {
        case kind
        case action
        case value
        case direction
        case pages
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch try container.decode(Kind.self, forKey: .kind) {
        case .action:
            self = .action(ComputerUseAXActionEvent(
                action: try container.decode(ComputerUseAXElementAction.self, forKey: .action)
            ))
        case .setValue:
            self = .setValue(ComputerUseAXSetValueEvent(
                value: try container.decode(String.self, forKey: .value)
            ))
        case .setSelectedText:
            self = .setSelectedText(ComputerUseAXSetValueEvent(
                value: try container.decode(String.self, forKey: .value)
            ))
        case .focus:
            self = .focus
        case .scroll:
            self = .scroll(ComputerUseAXScrollEvent(
                direction: try container.decode(ComputerUseAXScrollDirection.self, forKey: .direction),
                pages: try container.decode(Double.self, forKey: .pages)
            ))
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .action(let event):
            try container.encode(Kind.action, forKey: .kind)
            try container.encode(event.action, forKey: .action)
        case .setValue(let event):
            try container.encode(Kind.setValue, forKey: .kind)
            try container.encode(event.value, forKey: .value)
        case .setSelectedText(let event):
            try container.encode(Kind.setSelectedText, forKey: .kind)
            try container.encode(event.value, forKey: .value)
        case .focus:
            try container.encode(Kind.focus, forKey: .kind)
        case .scroll(let event):
            try container.encode(Kind.scroll, forKey: .kind)
            try container.encode(event.direction, forKey: .direction)
            try container.encode(event.pages, forKey: .pages)
        }
    }
}

public struct ComputerUsePostEventToAXElementParams: Codable, Sendable, Equatable {
    public let windowId: Int
    public let stateId: String
    public let elementIndex: Int
    public let event: ComputerUseAXElementEvent

    public init(
        windowId: Int,
        stateId: String,
        elementIndex: Int,
        event: ComputerUseAXElementEvent
    ) {
        self.windowId = windowId
        self.stateId = stateId
        self.elementIndex = elementIndex
        self.event = event
    }
}

public struct ComputerUsePostEventToAXElementResult: Codable, Sendable, Equatable {
    public let pid: Int
    public let windowId: Int
    public let stateId: String
    public let elementIndex: Int
    public let event: ComputerUseAXElementEvent

    public init(
        pid: Int,
        windowId: Int,
        stateId: String,
        elementIndex: Int,
        event: ComputerUseAXElementEvent
    ) {
        self.pid = pid
        self.windowId = windowId
        self.stateId = stateId
        self.elementIndex = elementIndex
        self.event = event
    }
}

File: sidecar/src/agent/tools/builtins/computer-use.ts
import { readFileSync, unlinkSync } from "node:fs";
import { z, type ZodType } from "zod";
import { supportsVision } from "../../../llm/models/capabilities";
import type { Tool, ToolResultContent } from "../../../llm/types";
import type { Dispatcher } from "../../../rpc/dispatcher";
import {
  RPCMethod,
  type ComputerUseBounds,
  type ComputerUseGetAppStateParams,
  type ComputerUseGetAppStateResult,
  type ComputerUseListAppsParams,
  type ComputerUseListWindowsParams,
  type ComputerUsePostEventToAXElementParams,
  type ComputerUsePostKeyboardEventParams,
  type ComputerUsePostMouseEventParams,
  type ComputerUseStartAppSessionParams,
} from "../../../rpc/rpc-types";
import { defineTool, zodParametersToJSONSchema } from "../core/schema";
import type { ToolRegistry } from "../core/registry";
import {
  ToolUserError,
  type ToolExecContext,
  type ToolExecResult,
  type ToolHandler,
  type ToolRuntimeEffects,
} from "../core/types";

export type ComputerUseStopAppSessionArgs = Record<string, never>;

export type ComputerUseArgs =
  | ComputerUseListAppsParams
  | ComputerUseListWindowsParams
  | ComputerUseGetAppStateParams
  | ComputerUseStartAppSessionParams
  | ComputerUseStopAppSessionArgs
  | ComputerUsePostMouseEventParams
  | ComputerUsePostKeyboardEventParams
  | ComputerUsePostEventToAXElementParams;

const pointSchema = z.object({
  x: z.number(),
  y: z.number(),
}).strict();

const modifierSchema = z.enum(["command", "shift", "option", "control", "function"]);

const mouseEventSchema = z.discriminatedUnion("kind", [
  z.object({
    kind: z.literal("click"),
    button: z.enum(["left", "right"]),
    point: pointSchema,
    count: z.number().int().optional(),
  }).strict(),
  z.object({
    kind: z.literal("drag"),
    button: z.enum(["left", "right"]),
    from: pointSchema,
    to: pointSchema,
  }).strict(),
]);

const keyboardEventSchema = z.discriminatedUnion("kind", [
  z.object({
    kind: z.literal("text"),
    text: z.string(),
    delayMilliseconds: z.number().int().optional(),
  }).strict(),
  z.object({
    kind: z.literal("keyPress"),
    key: z.string(),
    modifiers: z.array(modifierSchema).optional(),
    count: z.number().int().optional(),
  }).strict(),
  z.object({
    kind: z.literal("hotkey"),
    modifiers: z.array(modifierSchema),
    key: z.string(),
  }).strict(),
]);

const axEventSchema = z.discriminatedUnion("kind", [
  z.object({
    kind: z.literal("focus"),
  }).strict(),
  z.object({
    kind: z.literal("action"),
    action: z.enum([
      "press",
      "showMenu",
      "pick",
      "confirm",
      "cancel",
      "open",
      "increment",
      "decrement",
      "scrollToVisible",
    ]),
  }).strict(),
  z.object({
    kind: z.literal("setValue"),
    value: z.string(),
  }).strict(),
  z.object({
    kind: z.literal("setSelectedText"),
    value: z.string(),
  }).strict(),
  z.object({
    kind: z.literal("scroll"),
    direction: z.enum(["up", "down", "left", "right"]),
    pages: z.number(),
  }).strict(),
]);

export const COMPUTER_USE_TOOL_NAMES = [
  "list_apps",
  "list_windows",
  "get_app_state",
  "start_app_session",
  "stop_app_session",
  "use_mouse",
  "use_keyboard",
  "perform_AX_action",
] as const;

export type ComputerUseToolName = (typeof COMPUTER_USE_TOOL_NAMES)[number];

export type ComputerUseToolSpec = Tool & {
  name: ComputerUseToolName;
};

interface ComputerUseToolDefinition {
  name: ComputerUseToolName;
  description: string;
  parameters: ZodType<ComputerUseArgs, any, any>;
}

interface ComputerUseRpcToolBinding {
  method: string;
  prepare?: (args: ComputerUseArgs, ctx: ToolExecContext) => object;
  render?: (result: unknown, ctx: ToolExecContext) => ToolResultContent[];
  applyRuntimeEffects?: (
    result: ToolExecResult<unknown>,
    args: ComputerUseArgs,
    ctx: ToolExecContext,
    effects: ToolRuntimeEffects,
  ) => void;
}

const COMPUTER_USE_TOOL_DEFINITIONS: ComputerUseToolDefinition[] = [
  {
    name: "list_apps",
    description: "List macOS applications available to Notch Agent Computer Use.",
    parameters: z.object({
      mode: z.enum(["running", "all"]).describe("Use running unless installed apps are needed."),
    }).strict(),
  },
  {
    name: "list_windows",
    description: "List windows for a running macOS app pid.",
    parameters: z.object({
      pid: z.number().int().describe("Target app process id."),
    }).strict(),
  },
  {
    name: "get_app_state",
    description:
      "Read a target window's AX tree and optionally a screenshot in the active Computer Use app session. Returns stateId for AX element actions and screenshot-local mouse actions.",
    parameters: z.object({
      windowId: z.number().int(),
      captureMode: z.enum(["vision", "ax"]),
    }).strict(),
  },
  {
    name: "start_app_session",
    description: "Start a Computer Use app session for app state reads and mouse, keyboard, and AX element actions.",
    parameters: z.object({
      pid: z.number().int(),
      windowId: z.number().int(),
    }).strict(),
  },
  {
    name: "stop_app_session",
    description: "Stop the active Computer Use app session and run cleanup.",
    parameters: z.object({}).strict(),
  },
  {
    name: "use_mouse",
    description:
      "Post a mouse event to a window in the active Computer Use app session. " +
      "Use the stateId from the get_app_state screenshot you are pointing at. " +
      "event variants: " +
      "{kind:\"click\", button:\"left\"|\"right\", point:{x:number,y:number}, count?:integer}; " +
      "{kind:\"drag\", button:\"left\"|\"right\", from:{x:number,y:number}, to:{x:number,y:number}}. " +
      "Coordinates are screenshot-local pixels in that get_app_state image.",
    parameters: z.object({
      windowId: z.number().int(),
      stateId: z.string().describe("State ID from the get_app_state screenshot used for these coordinates."),
      event: mouseEventSchema.describe(
        "Mouse event: {kind:'click', button:'left'|'right', point:{x,y}, count?} or {kind:'drag', button:'left'|'right', from:{x,y}, to:{x,y}}. Coordinates are screenshot-local pixels from the referenced get_app_state image.",
      ),
    }).strict(),
  },
  {
    name: "use_keyboard",
    description:
      "Post a keyboard event to a window in the active Computer Use app session. " +
      "event variants: " +
      "{kind:\"text\", text:string, delayMilliseconds?:integer}; " +
      "{kind:\"keyPress\", key:string, modifiers?:modifier[], count?:integer}; " +
      "{kind:\"hotkey\", modifiers:modifier[], key:string}. " +
      "modifier = \"command\"|\"shift\"|\"option\"|\"control\"|\"function\".",
    parameters: z.object({
      windowId: z.number().int(),
      event: keyboardEventSchema.describe(
        "Keyboard event: {kind:'text', text, delayMilliseconds?}, {kind:'keyPress', key, modifiers?, count?}, or {kind:'hotkey', modifiers, key}.",
      ),
    }).strict(),
  },
  {
    name: "perform_AX_action",
    description:
      "Perform a semantic AX event on an element from a prior get_app_state result. " +
      "event variants: " +
      "{kind:\"focus\"}; " +
      "{kind:\"action\", action:\"press\"|\"showMenu\"|\"pick\"|\"confirm\"|\"cancel\"|\"open\"|\"increment\"|\"decrement\"|\"scrollToVisible\"}; " +
      "{kind:\"setValue\", value:string}; " +
      "{kind:\"setSelectedText\", value:string}; " +
      "{kind:\"scroll\", direction:\"up\"|\"down\"|\"left\"|\"right\", pages:number}.",
    parameters: z.object({
      windowId: z.number().int(),
      stateId: z.string(),
      elementIndex: z.number().int(),
      event: axEventSchema.describe(
        "AX event: {kind:'focus'}, {kind:'action', action}, {kind:'setValue', value}, {kind:'setSelectedText', value}, or {kind:'scroll', direction, pages}.",
      ),
    }).strict(),
  },
];

const COMPUTER_USE_TOOL_BINDINGS = {
  list_apps: {
    method: RPCMethod.computerUseListApps,
  },
  list_windows: {
    method: RPCMethod.computerUseListWindows,
  },
  get_app_state: {
    method: RPCMethod.computerUseGetAppState,
    render: renderComputerUseAppStateResult,
  },
  start_app_session: {
    method: RPCMethod.computerUseStartAppSession,
    applyRuntimeEffects: recordStartedAppSession,
  },
  stop_app_session: {
    method: RPCMethod.computerUseStopAppSession,
    prepare: (_args, ctx) => stopAppSessionParams(ctx),
    applyRuntimeEffects: clearStartedAppSession,
  },
  use_mouse: {
    method: RPCMethod.computerUsePostMouseEvent,
  },
  use_keyboard: {
    method: RPCMethod.computerUsePostKeyboardEvent,
  },
  perform_AX_action: {
    method: RPCMethod.computerUsePostEventToAXElement,
  },
} satisfies Record<ComputerUseToolName, ComputerUseRpcToolBinding>;

export const COMPUTER_USE_TOOL_SPECS: ComputerUseToolSpec[] = COMPUTER_USE_TOOL_DEFINITIONS.map((tool) => ({
  name: tool.name,
  description: tool.description,
  parameters: zodParametersToJSONSchema(tool.parameters),
}));

export function registerComputerUseTools(registry: ToolRegistry, dispatcher: Dispatcher): void {
  for (const tool of createComputerUseTools(dispatcher)) {
    registry.register(tool);
  }
}

export function createComputerUseTools(dispatcher: Dispatcher): ToolHandler<ComputerUseArgs, unknown>[] {
  return COMPUTER_USE_TOOL_DEFINITIONS.map((definition) =>
    createComputerUseRpcTool(definition, COMPUTER_USE_TOOL_BINDINGS[definition.name], dispatcher)
  );
}

function createComputerUseRpcTool(
  definition: ComputerUseToolDefinition,
  binding: ComputerUseRpcToolBinding,
  dispatcher: Dispatcher,
): ToolHandler<ComputerUseArgs, unknown> {
  return defineTool({
    name: definition.name,
    description: definition.description,
    parameters: definition.parameters,
    execute: async (args, ctx) => {
      const rpcArgs = binding.prepare ? binding.prepare(args, ctx) : args;
      const result = await dispatcher.request(binding.method, rpcArgs, { signal: ctx.signal });
      return {
        content: binding.render ? binding.render(result, ctx) : [{ type: "text", text: JSON.stringify(result, null, 2) }],
        details: result,
        isError: false,
      } satisfies ToolExecResult;
    },
    applyRuntimeEffects: binding.applyRuntimeEffects,
  });
}

function stopAppSessionParams(ctx: ToolExecContext): { pid: number } {
  const appSession = ctx.computerUse?.appSession;
  if (!appSession) {
    throw new ToolUserError("no app session was started by this agent session.");
  }
  return { pid: appSession.pid };
}

function recordStartedAppSession(
  _result: ToolExecResult<unknown>,
  args: ComputerUseArgs,
  _ctx: ToolExecContext,
  effects: ToolRuntimeEffects,
): void {
  const { pid, windowId } = args as ComputerUseStartAppSessionParams;
  effects.computerUse.setAppSession({
    pid,
    windowId,
  });
}

function clearStartedAppSession(
  _result: ToolExecResult<unknown>,
  _args: ComputerUseArgs,
  _ctx: ToolExecContext,
  effects: ToolRuntimeEffects,
): void {
  effects.computerUse.clearAppSession();
}

function renderComputerUseAppStateResult(result: unknown, ctx: ToolExecContext): ToolResultContent[] {
  const state = result as ComputerUseGetAppStateResult;
  const screenshot = state.screenshot;
  const content: ToolResultContent[] = [{ type: "text", text: renderComputerUseAppStateText(state) }];
  if (screenshot) {
    const imageContent = consumeScreenshotFile(screenshot, ctx);
    if (imageContent) content.push(imageContent);
  }
  return content;
}

export function renderComputerUseAppStateText(state: ComputerUseGetAppStateResult): string {
  const lines = [
    "<app_state>",
    `App=${renderAppIdentity(state)} (pid ${state.pid})`,
    `State ID: ${state.stateId}`,
    `Elements: ${state.elementCount}`,
  ];
  if (state.screenshot) {
    lines.push(
      `Screenshot: ${state.screenshot.format} ${state.screenshot.width}x${state.screenshot.height} px @${state.screenshot.scaleFactor}x`,
      `Screenshot windowFrame: ${renderBounds(state.screenshot.coordinateSpace.windowFrame)}`,
      `Screenshot windowBounds: ${renderBounds(state.screenshot.coordinateSpace.windowBounds)}`,
      `Screenshot pixelSize: ${state.screenshot.coordinateSpace.pixelSize.width}x${state.screenshot.coordinateSpace.pixelSize.height} px`,
    );
  }
  const tree = state.treeMarkdown.trim();
  if (tree.length > 0) lines.push(tree);
  lines.push("</app_state>");
  return lines.join("\n");
}

function consumeScreenshotFile(
  screenshot: NonNullable<ComputerUseGetAppStateResult["screenshot"]>,
  ctx: ToolExecContext,
): ToolResultContent | null {
  try {
    if (!supportsVision(ctx.model)) return null;
    return {
      type: "image",
      data: readFileSync(screenshot.imagePath).toString("base64"),
      mimeType: screenshot.format === "jpeg" ? "image/jpeg" : "image/png",
    };
  } finally {
    unlinkSync(screenshot.imagePath);
  }
}

function renderBounds(bounds: ComputerUseBounds): string {
  return `x=${bounds.x} y=${bounds.y} width=${bounds.width} height=${bounds.height}`;
}

function renderAppIdentity(state: ComputerUseGetAppStateResult): string {
  if (state.bundleId) return state.bundleId;
  if (state.appName) return state.appName;
  return "unknown";
}
